### PR TITLE
Add ClassDiscovery::safeClassExists

### LINKED ISSUE #107: Composer dependency missing?
| Q            | A
| ------------ | ---
| Bug?         | no
| New Feature? | no
| Version      | master

Just a quick question: In `Http\Discovery\Strategy\CommonClassesStrategy` you use classes i.e. `Nyholm\Psr7\Request` and this dependency is not defined in the `composer.json`. How does it work? This package composer.json states:
```
"require": {
        "php": "^5.5 || ^7.0"
    }
```

The Nyholm composer.json (https://github.com/Nyholm/psr7/blob/master/composer.json) states:
```
"require": {
        "php": "^7.1",
```

How can I use this package without conflicts and how is it supposed to be used when the dependencies are not explicitly declared? Thanks a lot guys.

### PR DESCRIPTION
**Pull request by Nyholm**
| Q               | A
| --------------- | ---
| Bug fix?        | yes
| New feature?    | no
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | fixes #107
| Documentation   | 
| License         | MIT

#### What's in this PR?

This is an alternative to #110.

### CODE REVIEW

**Comment by Nyholm** [comment]
@localheinz could you try this patch? Does it work for you?

**Comment by localheinz** [comment]
@dbu @Nyholm 

Will do, give me a moment!

**Comment by localheinz** [comment]
Unfortunately fails with

```
Warning: include(Nyholm\Psr7\Request.php): failed to open stream: No such file or directory  in /www/shop/lib/Varien/Autoload.php on line 94
```

Personally, though, I have worked around the issue by creating the client without using `php-http/discovery`. Works fine for my use case (legacy application), but might not for others.

**Comment by Nyholm** [comment]
That is strange.. I tried to reproduce and the patch do work.

**Comment by localheinz** [comment]
@Nyholm 

Maybe a cache issue, let me try again!

**Comment by localheinz** [comment]
@Nyholm 

No, issue still persists. But don't worry, I assume this Magento installation (`1.9.1.2`) has been patched a bit, maybe that's an issue. 

```
Warning: include(Nyholm\Psr7\Request.php): failed to open stream: No such file or directory  in /www/shop/lib/Varien/Autoload.php on line 94

#0 /www/shop/lib/Varien/Autoload.php(94): mageCoreErrorHandler(2, 'include(Nyholm\\...', '/www/shop/lib/V...', 94, Array)
#1 /www/shop/lib/Varien/Autoload.php(94): include()
#2 [internal function]: Varien_Autoload->autoload('Nyholm\\Psr7\\Req...')
#3 [internal function]: spl_autoload_call('Nyholm\\Psr7\\Req...')
#4 /www/shop/vendor/php-http/discovery/src/ClassDiscovery.php(165): is_callable(Array)
#5 /www/shop/vendor/php-http/discovery/src/ClassDiscovery.php(63): Http\Discovery\ClassDiscovery::evaluateCondition(Array)
#6 /www/shop/vendor/php-http/discovery/src/MessageFactoryDiscovery.php(25): Http\Discovery\ClassDiscovery::findOneByType('Http\\Message\\Me...')
#7 /www/shop/vendor/foo/bar-api/src/Client.php(104): Http\Discovery\MessageFactoryDiscovery::find()
#8 /www/shop/app/code/local/Foo/Bar/Helper/Externalservice.php(37): Foo\Bar\Api\Client::create(Object(Http\Adapter\Guzzle6\Client))
#9 /www/shop/app/Mage.php(557): Foo_Bar_Helper_Externalservice->__construct()
#10 /www/shop/app/code/local/Foo/Bar/controllers/CategoryController.php(29): Mage::helper('coursemanager/e...')
#11 /www/shop/app/code/local/Foo/Bar/controllers/CategoryController.php(19): Foo_Bar_CategoryController->api()
#12 /www/shop/app/code/core/Mage/Core/Controller/Varien/Action.php(418): Foo_Bar_CategoryController->listAction()
#13 /www/shop/app/code/core/Mage/Core/Controller/Varien/Router/Standard.php(254): Mage_Core_Controller_Varien_Action->dispatch('list')
#14 /www/shop/app/code/core/Mage/Core/Controller/Varien/Front.php(172): Mage_Core_Controller_Varien_Router_Standard->match(Object(Mage_Core_Controller_Request_Http))
#15 /www/shop/app/code/core/Mage/Core/Model/App.php(365): Mage_Core_Controller_Varien_Front->dispatch()
#16 /www/shop/app/Mage.php(693): Mage_Core_Model_App->run(Array)
#17 /www/shop/index.php(94): Mage::run('default', 'store')
#18 {main}
```

**Comment by Nyholm** [comment]
Okey. I cloned https://github.com/magento/magento2. Tested with last commit (which is the same as  version 2.2.5). 

Should we consider this as an accepted solution or should we try to target Magento 1 as well?

**Comment by dbu** [comment]
should we merge this?

**Comment by Nyholm** [comment]
Yes please
